### ISSUE TITLE
VxScan: no CVR saved on polls closed

### ISSUE THREAD

**Issue by mattroe** [votingworks/vxsuite#1981]
This issue was observed during L&A in Warren and reproduced on a locked-down production imaged VxScan.

CVR files are not saved to USB disk after polls are closed, but they are still saved if you either 1) power down the machine before ejection or 2) export CVRs through the admin card screen.

### Repro steps
1. Scan ballots
2. Insert poll worker card and close polls
3. Remove USB
4. No CVR files will be saved

**Comment by benadida** [comment]
data from the field: I don't think we eject the disk properly. I think the operation gets interrupted. If we take out the USB stick before shutting down vxscan, there is no data on it. But if we shut it down, which forces an eject/flush, then the data is there.

**Comment by adghayes** [comment]
@mattroe @benadida - this is an intermittent issue, correct? It seems like once we get it on a machine it is reproducible, but it is not reproducible on all machines, all the time. I played with this most of today and learned this can be reproduced in a VM with hardware, but I have not found a way to do it consistently or frequently.

I do think that @benadida is right in that data is failing to get flushed to disk sometimes. I'm not sure if this is related, but I few times I caught the cast-vote-record directory on the USB drive looking like this:

![Image](https://user-images.githubusercontent.com/37960853/180106255-93f87275-6fc2-4fdd-81f4-50b89cd63ae7.png)

![Image](https://user-images.githubusercontent.com/37960853/180106305-f661e257-dc62-4185-a374-06d39e394ded.png)

It was temporarily malformed. And I once saw it switch to to being normal again after this. Note that this was many seconds after our app UI had said the USB was ready.

@benadida my working theory is that we are doing our I/O right, but closing our write streams doesn't guarantee that data will be flushed to disk immediately or even soon. If that's the case, there are a couple ways we might be able to force that:

- Ejecting / unmounting the disk, which we currently don't do when exporting results from the precinct scanner during polls closed
- Forcibly `sync`ing the file by opening up a `FileHandle` and calling `sync` on it

**Comment by benadida** [comment]
Wait, we're not ejecting or syncing? Then yeah it definitely won't work.

**Comment by adghayes** [comment]
We're not ejecting on polls closed, nope. We're not going out of our way to sync in any of the apps.

In the case of the VxScan not exporting CVRs (this ticket), the appropriate fix might be just ejecting after we save the CVR to file?

Any other place in the code base where we don't offer an option for the user to eject (exporting logs), or users don't click Eject, in theory the same thing fail to flush could happen. Do you think we should go out of our way to sync all files after closing write streams with something like this:

``` 
export async function sync(path: string | Buffer) {
  const fileHandle = await nodeOpen(path, 'r');
  await fileHandle.sync();
  await fileHandle.close();
}
```

In either case, I'll have trouble feeling good about closing this because of the trouble I'm having repro'ing.

**Comment by adghayes** [comment]
@benadida and I discussed and agreed that, as a next step, we would try mounting devices with  the `-sync` flag so that there is no write caching to later flush with a `sync` or  an unmount. 

I have a [kiosk-browser branch](https://github.com/votingworks/kiosk-browser/tree/mount-sync) up with the change. I've done two tests of the new process:

On VxScan VxDev, I tested closing polls and checking whether the file was successfully exported:
- On `main`, the file was successfully exported 0 of 5 times
- On `mount-sync`, the file was successfully exported 5 of 5 times

On a Debian machine (VxDev), I tested downloading a 1 GB to a wimpy USB drive. I started an unmount command immediately after the download began and timed how long it took to unmount the drive (when all data was presumable flushed).
- Mounted without the `-sync` flag, the download took 2 minutes, 7 seconds
- Mounted with the `-sync` flag, the download took 2 minutes, 10 seconds

The difference could be just varying time it took me to notice.

@eventualbuddha @benadida what other testing do you think we should do?

**Comment by eventualbuddha** [comment]
I'd want to test the full stack of downloading ~1GB through scan service through kiosk-browser to the USB. You don't need to actually build up a backup that big, maybe just hardcode an endpoint to stream that same 1GB file?

**Comment by benadida** [comment]
@adghayes fascinating, and just wondering what are you comparing to? You wrote both the sync mount and separate flush approach and compared those two?

**Comment by adghayes** [comment]
@benadida I used VxDev. In the transfer tests, I would mount the USB in sync or async mode, via CL or via our software (from the UI). Then I would initialize a file transfer, through nautilus or our software. Then I would open up another terminal and keep trying to unmount, and measure the time it took from initiating the download to allowing me to unmount.

In sync mode, nautilus was also giving accurate download time estimates. 

I think the initial test I did was bad because I hadn’t turned off automounting yet and it mounted in async mode both times
